src/rubyfmt.rb

```diff
@@ -1824,26 +1813,53 @@ def format_program(line_metadata, sexp, result)
 end
 
 def extract_line_metadata(file_data)
-  metadata = Struct.new(:comment_blocks, :w_lines)
+  metadata = Struct.new(:comment_blocks)
   comment_blocks = {}
-  w_lines = {}
+
   file_data.split("\n").each_with_index do |line, index|
-    if /^ *#/ === line
-      comment_blocks[index] = line
-    elsif /%w/i === line
-      w_lines[index+1] = line
-    end
+    comment_blocks[index] = line if /^ *#/ === line
   end
 
-  metadata.new(comment_blocks, w_lines)
+  metadata.new(comment_blocks)
+end
+
+class Parser < Ripper::SexpBuilderPP
+  ARRAY_SYMBOLS = {
+    qsymbols: '%i',
+    qwords: '%w',
+    symbols: '%I',
+    words: '%W'
+  }.freeze
+
+  def self.is_percent_array?(rest)
+    ARRAY_SYMBOLS.include?(rest[0][0])
+  end
+
+  def self.percent_symbol_for(rest)
+    ARRAY_SYMBOLS[rest[0][0]]
+  end
+
+  private
+
+  ARRAY_SYMBOLS.each do |event, symbol|
+    define_method(:"on_#{event}_new") do
+      [event, [], [lineno, column]]
+    end
+
+    define_method(:"on_#{event}_add") do |parts, part|
+      parts.tap do |node|
+        node[1] << part
+      end
+    end
+  end
 end
 
 def main
   file_data = File.read(FILE)
   file_data = file_data.gsub("\r\n", "\n")
 
   line_metadata = extract_line_metadata(file_data)
-  sexp = Ripper.sexp(file_data)
+  sexp = Parser.new(file_data).parse
```

Comment:
Heads up because I just ran into this with prettier, make sure to check `parser.error?` so that you can bail out if ripper failed to generate the AST